File: server/lib/supervisor.ts
/**
 * Databricks Supervisor Agent — VR Dashboard pattern.
 * Chart data from SQL cache; AI tab summaries via Supervisor → Genie (MAS endpoint).
 */
import type { MetricsPayload } from '../../shared/types/dashboard.js';
import { buildTabInsights } from './metricsTransform.js';
import { buildSupervisorPrompt, type SummaryEntity } from './summaryPrompts.js';

type Message = { role: string; content: string };
type SupervisorInput = Message | Record<string, unknown>;

const DEBUG = (process.env.SUPERVISOR_DEBUG || 'false').toLowerCase() === 'true';
const MAX_CONTINUATIONS = Math.max(1, Number(process.env.SUPERVISOR_MAX_CONTINUATIONS || 12));
const LONG_TASK = String(process.env.SUPERVISOR_LONG_TASK ?? 'true').toLowerCase() !== 'false';

function host(): string {
  return (process.env.DATABRICKS_HOST || process.env.DATABRICKS_SERVER_HOSTNAME || '')
    .replace(/^https?:\/\//, '')
    .replace(/\/$/, '');
}

function token(): string {
  return process.env.DATABRICKS_PAT_TOKEN || process.env.DATABRICKS_TOKEN || '';
}

function endpoint(): string {
  return process.env.SUPERVISOR_ENDPOINT_NAME || '';
}

export function supervisorConfigured(): boolean {
  return Boolean(host() && token() && endpoint());
}

function supervisorDatabricksOptions(): Record<string, unknown> {
  const opts: Record<string, unknown> = {};
  if (LONG_TASK) opts.long_task = true;
  return opts;
}

function isTimeoutResponse(text: string): boolean {
  const lower = text.toLowerCase();
  return (
    (lower.includes('time out') || lower.includes('time limit') || lower.includes('timed out')) &&
    (lower.includes('continue') || lower.includes('would you like') || lower.includes('shall i'))
  );
}

function findTaskContinueCheckpoint(raw: string): { id: string; step?: number } | null {
  const idMatch = raw.match(/"type"\s*:\s*"task_continue_request"[\s\S]*?"id"\s*:\s*"(continue_[^"]+)"/);
  if (!idMatch) return null;
  const stepMatch = raw.match(/"type"\s*:\s*"task_continue_request"[\s\S]*?"step"\s*:\s*(\d+)/);
  return { id: idMatch[1], step: stepMatch ? Number(stepMatch[1]) : undefined };
}

function parseJsonBlock(raw: string): Record<string, unknown> {
  const text = raw.trim();
  try {
    const d = JSON.parse(text) as unknown;
    if (d && typeof d === 'object') return unwrapResponse(d as Record<string, unknown>);
  } catch {
    // continue
  }

  const cleaned = text.replace(/```json\s*|```\s*/g, '').trim();
  let best: Record<string, unknown> | null = null;
  let bestLen = 0;
  let depth = 0;
  let start: number | null = null;

  for (let i = 0; i < cleaned.length; i++) {
    const ch = cleaned[i];
    if (ch === '{') {
      if (depth === 0) start = i;
      depth++;
    } else if (ch === '}') {
      depth--;
      if (depth === 0 && start != null) {
        const candidate = cleaned.slice(start, i + 1);
        if (candidate.length > bestLen) {
          try {
            const d = JSON.parse(candidate) as Record<string, unknown>;
            if (d && Object.keys(d).length) {
              best = d;
              bestLen = candidate.length;
            }
          } catch {
            // skip
          }
        }
      }
    }
  }

  if (best) return unwrapResponse(best);
  if (looksLikeSseStream(cleaned)) return {};
  return { narrative: cleaned.slice(0, 500) };
}

function unwrapResponse(d: Record<string, unknown>): Record<string, unknown> {
  if (Array.isArray(d.output)) {
    const allTexts: string[] = [];
    for (const block of d.output) {
      if (!block || typeof block !== 'object') continue;
      const b = block as Record<string, unknown>;
      if (b.type === 'message' && Array.isArray(b.content)) {
        for (const cb of b.content) {
          if (cb && typeof cb === 'object') {
            const c = cb as Record<string, unknown>;
            if (c.type === 'output_text' || c.type === 'text') {
              const t = String(c.text || '').trim();
              if (t) allTexts.push(t);
            }
          }
        }
      } else if (b.type === 'output_text' || b.type === 'text') {
        const t = String(b.text || '').trim();
        if (t) allTexts.push(t);
      }
    }

    if (DEBUG) console.log(`[supervisor] found ${allTexts.length} text block(s) in output`);

    if (allTexts.length) {
      for (const text of [...allTexts].reverse()) {
        if (text.includes('{') && (text.includes('narrative') || text.includes('key_insights') || text.includes('insight'))) {
          return parseJsonBlock(text);
        }
      }

      const combined = allTexts.join(' ');
      if (combined.includes('{')) {
        const result = parseJsonBlock(combined);
        if (typeof result.narrative === 'string' || result.key_insights || result.insight) {
          return result;
        }
      }

      return parseJsonBlock(allTexts[allTexts.length - 1]);
    }

    return { narrative: 'Supervisor is still querying Genie — please retry in a moment.' };
  }

  if (Array.isArray(d.choices)) {
    const c = d.choices[0] as Record<string, unknown>;
    const msg = (c?.message as Record<string, unknown>) || {};
    const content = msg.content;
    if (typeof content === 'string') return parseJsonBlock(content);
    if (content && typeof content === 'object') return content as Record<string, unknown>;
  }

  if (typeof d.output === 'string') return parseJsonBlock(d.output);

  if ('narrative' in d || 'key_insights' in d || 'observations' in d || 'insight' in d) return d;

  return d;
}

function extractSseDataPayloads(raw: string): string[] {
  const payloads: string[] = [];
  const re = /^data:\s*(.*)$/gm;
  let match: RegExpExecArray | null;
  while ((match = re.exec(raw)) !== null) {
    payloads.push(match[1].trim());
  }
  return payloads;
}

function looksLikeSseStream(raw: string): boolean {
  return raw.includes('data:') && (
    raw.includes('response.output_text.delta') ||
    raw.includes('response.completed') ||
    raw.includes('response.output_item.done')
  );
}

/** Reassemble Databricks Open Responses SSE chunks into plain text or a JSON response body. */
function parseSupervisorSseStream(raw: string): { text: string; rawForContinue: string } {
  if (!looksLikeSseStream(raw)) {
    return { text: raw, rawForContinue: raw };
  }

  const deltas: string[] = [];
  const outputTexts: string[] = [];
  let completedResponse: Record<string, unknown> | null = null;
  const continueLines: string[] = [];

  for (const dataLine of extractSseDataPayloads(raw)) {
    if (!dataLine || dataLine === '[DONE]') continue;

    if (dataLine.includes('task_continue_request')) {
      continueLines.push(`data: ${dataLine}`);
    }

    try {
      const event = JSON.parse(dataLine) as Record<string, unknown>;
      const type = String(event.type || '');

      if (type === 'response.output_text.delta') {
        deltas.push(String(event.delta || ''));
        continue;
      }

      if (type === 'response.completed' && event.response && typeof event.response === 'object') {
        completedResponse = event.response as Record<string, unknown>;
        continue;
      }

      if (type === 'response.output_item.done') {
        const item = event.item as Record<string, unknown> | undefined;
        if (Array.isArray(item?.content)) {
          for (const block of item.content) {
            if (!block || typeof block !== 'object') continue;
            const chunk = block as Record<string, unknown>;
            if (chunk.type === 'output_text' || chunk.type === 'text') {
              const t = String(chunk.text || '').trim();
              if (t) outputTexts.push(t);
            }
          }
        } else if (typeof item?.text === 'string' && item.text.trim()) {
          outputTexts.push(item.text.trim());
        }
      }
    } catch {
      // skip malformed SSE JSON lines
    }
  }

  const rawForContinue = continueLines.length ? continueLines.join('\n\n') : raw;

  if (completedResponse) {
    return { text: JSON.stringify(completedResponse), rawForContinue };
  }

  const assembled = (outputTexts.join('') || deltas.join('')).trim();
  if (assembled) {
    return { text: assembled, rawForContinue };
  }

  return { text: raw, rawForContinue };
}

async function readResponseBody(resp: Response): Promise<string> {
  if (!resp.body) return resp.text();

  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  const chunks: string[] = [];

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    const chunk = decoder.decode(value, { stream: true });
    chunks.push(chunk);
    if (DEBUG) console.log(`[supervisor] chunk: ${chunk.slice(0, 150)}`);
  }

  return chunks.join('');
}

async function postOnce(input: SupervisorInput[]): Promise<string> {
  const h = host();
  const url = `https://${h}/serving-endpoints/${endpoint()}/invocations`;
  const databricksOptions = supervisorDatabricksOptions();
  console.log(
    `[supervisor] -> ${endpoint()} | stream=true long_task=${Boolean(databricksOptions.long_task)}`,
  );

  const body: Record<string, unknown> = {
    input,
    stream: true,
  };
  if (Object.keys(databricksOptions).length) {
    body.databricks_options = databricksOptions;
  }

  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token()}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  console.log(`[supervisor] <- HTTP ${resp.status}`);
  if (!resp.ok) {
    const text = await resp.text();
    throw new Error(`Supervisor HTTP ${resp.status}: ${text.slice(0, 400)}`);
  }

  return readResponseBody(resp);
}

async function callSupervisor(messages: Message[], metrics?: MetricsPayload | null): Promise<string> {
  if (!supervisorConfigured()) {
    throw new Error('Supervisor not configured — set SUPERVISOR_ENDPOINT_NAME in .env');
  }

  const conversation: SupervisorInput[] = [...messages];
  let raw = '';

  for (let attempt = 0; attempt < MAX_CONTINUATIONS; attempt++) {
    console.log(`[supervisor] attempt ${attempt + 1}/${MAX_CONTINUATIONS}…`);
    const streamRaw = await postOnce(conversation);
    const { text: assembled, rawForContinue } = parseSupervisorSseStream(streamRaw);
    raw = assembled;
    if (DEBUG) console.log(`[supervisor] response ${raw.length} chars: ${raw.slice(0, 300)}`);

    const taskContinue = LONG_TASK ? findTaskContinueCheckpoint(rawForContinue) : null;
    if (taskContinue) {
      console.log(`[supervisor] task_continue checkpoint step=${taskContinue.step ?? '?'} — resuming…`);
      conversation.push({
        type: 'task_continue_request',
        id: taskContinue.id,
        ...(taskContinue.step != null ? { step: taskContinue.step } : {}),
      });
      conversation.push({
        type: 'task_continue_response',
        continue_request_id: taskContinue.id,
      });
      continue;
    }

    const parsed = parseJsonBlock(raw);
    const text = String(parsed.narrative || raw);

    if (isTimeoutResponse(text)) {
      console.log('[supervisor] ⚠ internal time limit — sending Continue…');
      conversation.push({ role: 'assistant', content: text });
      conversation.push({ role: 'user', content: 'Continue' });
      continue;
    }

    if (
      parsed.narrative &&
      Object.keys(parsed).length === 1 &&
      !text.includes('{') &&
      attempt < MAX_CONTINUATIONS - 1 &&
      text.length > 20
    ) {
      console.log('[supervisor] plain text — requesting JSON narrative…');
      conversation.push({ role: 'assistant', content: text });
      conversation.push({
        role: 'user',
        content: 'Continue and return the JSON with a narrative field now.',
      });
      continue;
    }

    if (parsed.narrative || parsed.key_insights || parsed.observations || parsed.insight) {
      return raw;
    }

    return raw;
  }

  throw new Error('Supervisor max continuations reached');
}

function isGarbageNarrative(text: string): boolean {
  const t = text.trim();
  return (
    looksLikeSseStream(t) ||
    t.startsWith('data:') ||
    t.includes('"response.output_text.delta"') ||
    t.includes('"type":"response.output_text.delta"')
  );
}

function extractNarrative(raw: string): string {
  const { text } = parseSupervisorSseStream(raw);
  const parsed = parseJsonBlock(text);

  if (typeof parsed.narrative === 'string' && parsed.narrative.trim()) {
    const narrative = parsed.narrative.trim();
    if (isGarbageNarrative(narrative)) return '';
    return narrative.slice(0, 600);
  }

  if (typeof parsed.insight === 'string' && parsed.insight.trim()) {
    return parsed.insight.trim().slice(0, 600);
  }

  if (Array.isArray(parsed.observations) && parsed.observations.length) {
    return String(parsed.observations[0]).slice(0, 600);
  }

  if (Array.isArray(parsed.key_insights)) {
    const first = parsed.key_insights[0] as Record<string, unknown> | undefined;
    if (first?.text) return String(first.text).slice(0, 600);
  }

  const cleaned = text.replace(/```json\s*|```\s*/g, '').trim();
  if (cleaned && !cleaned.startsWith('{') && !isGarbageNarrative(cleaned)) {
    return cleaned.slice(0, 600);
  }
  return '';
}

function narrativeUsesOnlyDashboardPeriods(narrative: string, periods: string[]): boolean {
  if (!periods.length) return true;
  const allowed = new Set(periods.map(p => p.toUpperCase()));
  const mentioned = narrative.match(/\bP\d{1,2}\b/gi) || [];
  return mentioned.every(m => allowed.has(m.toUpperCase()));
}

function metricsInsightFallback(
  entityType: SummaryEntity,
  metrics: MetricsPayload,
): string {
  return metrics.tab_insights?.[entityType] || buildTabInsights(metrics)[entityType] || '';
}

export async function getSupervisorSummary(
  entityType: SummaryEntity,
  filters: Record<string, unknown>,
  metrics?: MetricsPayload | null,
): Promise<{ narrative: string; source: 'supervisor' }> {
  const prompt = buildSupervisorPrompt(entityType, filters, metrics);
  console.log(`[supervisor] summary tab=${entityType} filters=${JSON.stringify(filters).slice(0, 120)}`);
  const raw = await callSupervisor([{ role: 'user', content: prompt }], metrics);
  let narrative = extractNarrative(raw);

  if (
    metrics?.periods?.length &&
    narrative &&
    !narrativeUsesOnlyDashboardPeriods(narrative, metrics.periods)
  ) {
    const fallback = metricsInsightFallback(entityType, metrics);
    if (fallback) {
      console.warn(
        `[supervisor] ${entityType} cited periods outside dashboard (${metrics.periods.join(', ')}) — using metrics insight`,
      );
      narrative = fallback;
    }
  }

  if (!narrative) {
    throw new Error('Supervisor returned empty narrative — retry shortly');
  }

  return { narrative, source: 'supervisor' };
}

/** VR parallel mode — one Supervisor call per tab, all run simultaneously. */
export async function getAllSupervisorSummaries(
  filters: Record<string, unknown>,
  metrics?: MetricsPayload | null,
): Promise<Record<SummaryEntity, string>> {
  const tabs: SummaryEntity[] = ['overview', 'category', 'line', 'dow', 'reason'];
  console.log(`[supervisor] PARALLEL mode — ${tabs.length} tab summaries (long_task=${LONG_TASK})`);

  const entries = await Promise.all(
    tabs.map(async tab => {
      try {
        const { narrative } = await getSupervisorSummary(tab, filters, metrics);
        return [tab, narrative] as const;
      } catch (e) {
        console.warn(`[supervisor] ${tab} failed:`, (e as Error).message);
        return [tab, ''] as const;
      }
    }),
  );

  return Object.fromEntries(entries) as Record<SummaryEntity, string>;
}

File: py_server/lib/supervisor.py
"""
Databricks Supervisor Agent — background job + Genie for tab summaries.
Chart data from SQL cache; AI tab summaries via Supervisor → Genie (MAS endpoint).
"""
from __future__ import annotations

import json
import os
import re
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any

from py_server.lib.databricks_fetch import databricks_fetch, databricks_host, databricks_token
from py_server.lib.metrics_transform import build_tab_insights
from py_server.lib.summary_prompts import SummaryEntity, build_supervisor_prompt

Message = dict[str, str]
SupervisorInput = Message | dict[str, Any]

DEBUG = (os.getenv('SUPERVISOR_DEBUG') or 'false').lower() == 'true'
MAX_CONTINUATIONS = max(1, int(os.getenv('SUPERVISOR_MAX_CONTINUATIONS') or 12))
LONG_TASK = (os.getenv('SUPERVISOR_LONG_TASK') or 'true').lower() != 'false'


def supervisor_configured() -> bool:
    return bool(databricks_host() and databricks_token() and _endpoint())


def _endpoint() -> str:
    return os.getenv('SUPERVISOR_ENDPOINT_NAME') or ''


def _supervisor_databricks_options() -> dict[str, Any]:
    opts: dict[str, Any] = {}
    if LONG_TASK:
        opts['long_task'] = True
    return opts


def _is_timeout_response(text: str) -> bool:
    lower = text.lower()
    return (
        ('time out' in lower or 'time limit' in lower or 'timed out' in lower)
        and ('continue' in lower or 'would you like' in lower or 'shall i' in lower)
    )


def _find_task_continue_checkpoint(raw: str) -> dict[str, Any] | None:
    id_match = re.search(
        r'"type"\s*:\s*"task_continue_request"[\s\S]*?"id"\s*:\s*"(continue_[^"]+)"',
        raw,
    )
    if not id_match:
        return None
    step_match = re.search(
        r'"type"\s*:\s*"task_continue_request"[\s\S]*?"step"\s*:\s*(\d+)',
        raw,
    )
    return {
        'id': id_match.group(1),
        'step': int(step_match.group(1)) if step_match else None,
    }


def _unwrap_response(d: dict[str, Any]) -> dict[str, Any]:
    output = d.get('output')
    if isinstance(output, list):
        all_texts: list[str] = []
        for block in output:
            if not isinstance(block, dict):
                continue
            if block.get('type') == 'message' and isinstance(block.get('content'), list):
                for cb in block['content']:
                    if isinstance(cb, dict) and cb.get('type') in ('output_text', 'text'):
                        t = str(cb.get('text') or '').strip()
                        if t:
                            all_texts.append(t)
            elif block.get('type') in ('output_text', 'text'):
                t = str(block.get('text') or '').strip()
                if t:
                    all_texts.append(t)

        if DEBUG:
            print(f'[supervisor] found {len(all_texts)} text block(s) in output', flush=True)

        if all_texts:
            for text in reversed(all_texts):
                if '{' in text and any(k in text for k in ('narrative', 'key_insights', 'insight')):
                    return _parse_json_block(text)

            combined = ' '.join(all_texts)
            if '{' in combined:
                result = _parse_json_block(combined)
                if isinstance(result.get('narrative'), str) or result.get('key_insights') or result.get('insight'):
                    return result

            return _parse_json_block(all_texts[-1])

        return {'narrative': 'Supervisor is still querying Genie — please retry in a moment.'}

    choices = d.get('choices')
    if isinstance(choices, list) and choices:
        c = choices[0] if isinstance(choices[0], dict) else {}
        msg = c.get('message') or {}
        content = msg.get('content')
        if isinstance(content, str):
            return _parse_json_block(content)
        if isinstance(content, dict):
            return content

    if isinstance(d.get('output'), str):
        return _parse_json_block(d['output'])

    if any(k in d for k in ('narrative', 'key_insights', 'observations', 'insight')):
        return d

    return d


def _parse_json_block(raw: str) -> dict[str, Any]:
    text = raw.strip()
    try:
        d = json.loads(text)
        if isinstance(d, dict):
            return _unwrap_response(d)
    except json.JSONDecodeError:
        pass

    cleaned = re.sub(r'```json\s*|```\s*', '', text).strip()
    best: dict[str, Any] | None = None
    best_len = 0
    depth = 0
    start: int | None = None

    for i, ch in enumerate(cleaned):
        if ch == '{':
            if depth == 0:
                start = i
            depth += 1
        elif ch == '}':
            depth -= 1
            if depth == 0 and start is not None:
                candidate = cleaned[start:i + 1]
                if len(candidate) > best_len:
                    try:
                        d = json.loads(candidate)
                        if isinstance(d, dict) and d:
                            best = d
                            best_len = len(candidate)
                    except json.JSONDecodeError:
                        pass

    if best:
        return _unwrap_response(best)
    if _looks_like_sse_stream(cleaned):
        return {}
    return {'narrative': cleaned[:500]}


def _extract_sse_data_payloads(raw: str) -> list[str]:
    return [m.group(1).strip() for m in re.finditer(r'^data:\s*(.*)$', raw, re.MULTILINE)]


def _looks_like_sse_stream(raw: str) -> bool:
    return 'data:' in raw and any(
        marker in raw
        for marker in (
            'response.output_text.delta',
            'response.completed',
            'response.output_item.done',
        )
    )


def _parse_supervisor_sse_stream(raw: str) -> tuple[str, str]:
    if not _looks_like_sse_stream(raw):
        return raw, raw

    deltas: list[str] = []
    output_texts: list[str] = []
    completed_response: dict[str, Any] | None = None
    continue_lines: list[str] = []

    for data_line in _extract_sse_data_payloads(raw):
        if not data_line or data_line == '[DONE]':
            continue

        if 'task_continue_request' in data_line:
            continue_lines.append(f'data: {data_line}')

        try:
            event = json.loads(data_line)
        except json.JSONDecodeError:
            continue

        etype = str(event.get('type') or '')
        if etype == 'response.output_text.delta':
            deltas.append(str(event.get('delta') or ''))
            continue

        if etype == 'response.completed' and isinstance(event.get('response'), dict):
            completed_response = event['response']
            continue

        if etype == 'response.output_item.done':
            item = event.get('item')
            if isinstance(item, dict):
                if isinstance(item.get('content'), list):
                    for block in item['content']:
                        if isinstance(block, dict) and block.get('type') in ('output_text', 'text'):
                            t = str(block.get('text') or '').strip()
                            if t:
                                output_texts.append(t)
                elif isinstance(item.get('text'), str) and item['text'].strip():
                    output_texts.append(item['text'].strip())

    raw_for_continue = '\n\n'.join(continue_lines) if continue_lines else raw

    if completed_response:
        return json.dumps(completed_response), raw_for_continue

    assembled = (''.join(output_texts) or ''.join(deltas)).strip()
    if assembled:
        return assembled, raw_for_continue

    return raw, raw_for_continue


def _read_response_body(resp: Any) -> str:
    chunks: list[str] = []
    for chunk in resp.iter_content(decode_unicode=True):
        if chunk:
            if DEBUG:
                print(f'[supervisor] chunk: {str(chunk)[:150]}', flush=True)
            chunks.append(chunk)
    return ''.join(chunks)


def _post_once(conversation: list[SupervisorInput]) -> str:
    h = databricks_host()
    url = f'https://{h}/serving-endpoints/{_endpoint()}/invocations'
    databricks_options = _supervisor_databricks_options()
    print(
        f'[supervisor] -> {_endpoint()} | stream=true long_task={bool(databricks_options.get("long_task"))}',
        flush=True,
    )

    body: dict[str, Any] = {'input': conversation, 'stream': True}
    if databricks_options:
        body['databricks_options'] = databricks_options

    resp = databricks_fetch(
        url,
        method='POST',
        headers={
            'Authorization': f'Bearer {databricks_token()}',
            'Content-Type': 'application/json',
        },
        json_body=body,
        stream=True,
    )

    print(f'[supervisor] <- HTTP {resp.status_code}', flush=True)
    if resp.status_code >= 400:
        text = resp.text[:400]
        raise RuntimeError(f'Supervisor HTTP {resp.status_code}: {text}')

    return _read_response_body(resp)


def _call_supervisor(messages: list[Message], metrics: dict[str, Any] | None = None) -> str:
    if not supervisor_configured():
        raise RuntimeError('Supervisor not configured — set SUPERVISOR_ENDPOINT_NAME in .env')

    conversation: list[SupervisorInput] = list(messages)
    raw = ''

    for attempt in range(MAX_CONTINUATIONS):
        print(f'[supervisor] attempt {attempt + 1}/{MAX_CONTINUATIONS}…', flush=True)
        stream_raw = _post_once(conversation)
        assembled, raw_for_continue = _parse_supervisor_sse_stream(stream_raw)
        raw = assembled
        if DEBUG:
            print(f'[supervisor] response {len(raw)} chars: {raw[:300]}', flush=True)

        task_continue = _find_task_continue_checkpoint(raw_for_continue) if LONG_TASK else None
        if task_continue:
            print(
                f'[supervisor] task_continue checkpoint step={task_continue.get("step", "?")} — resuming…',
                flush=True,
            )
            req: dict[str, Any] = {'type': 'task_continue_request', 'id': task_continue['id']}
            if task_continue.get('step') is not None:
                req['step'] = task_continue['step']
            conversation.append(req)
            conversation.append({
                'type': 'task_continue_response',
                'continue_request_id': task_continue['id'],
            })
            continue

        parsed = _parse_json_block(raw)
        text = str(parsed.get('narrative') or raw)

        if _is_timeout_response(text):
            print('[supervisor] ⚠ internal time limit — sending Continue…', flush=True)
            conversation.append({'role': 'assistant', 'content': text})
            conversation.append({'role': 'user', 'content': 'Continue'})
            continue

        if (
            parsed.get('narrative')
            and len(parsed) == 1
            and '{' not in text
            and attempt < MAX_CONTINUATIONS - 1
            and len(text) > 20
        ):
            print('[supervisor] plain text — requesting JSON narrative…', flush=True)
            conversation.append({'role': 'assistant', 'content': text})
            conversation.append({
                'role': 'user',
                'content': 'Continue and return the JSON with a narrative field now.',
            })
            continue

        if any(k in parsed for k in ('narrative', 'key_insights', 'observations', 'insight')):
            return raw

        return raw

    raise RuntimeError('Supervisor max continuations reached')


def _is_garbage_narrative(text: str) -> bool:
    t = text.strip()
    return (
        _looks_like_sse_stream(t)
        or t.startswith('data:')
        or 'response.output_text.delta' in t
        or '"type":"response.output_text.delta"' in t
    )


def _extract_narrative(raw: str) -> str:
    text, _ = _parse_supervisor_sse_stream(raw)
    parsed = _parse_json_block(text)

    narrative = parsed.get('narrative')
    if isinstance(narrative, str) and narrative.strip():
        n = narrative.strip()
        if _is_garbage_narrative(n):
            return ''
        return n[:600]

    insight = parsed.get('insight')
    if isinstance(insight, str) and insight.strip():
        return insight.strip()[:600]

    observations = parsed.get('observations')
    if isinstance(observations, list) and observations:
        return str(observations[0])[:600]

    key_insights = parsed.get('key_insights')
    if isinstance(key_insights, list) and key_insights:
        first = key_insights[0]
        if isinstance(first, dict) and first.get('text'):
            return str(first['text'])[:600]

    cleaned = re.sub(r'```json\s*|```\s*', '', text).strip()
    if cleaned and not cleaned.startswith('{') and not _is_garbage_narrative(cleaned):
        return cleaned[:600]
    return ''


def _narrative_uses_only_dashboard_periods(narrative: str, periods: list[str]) -> bool:
    if not periods:
        return True
    allowed = {p.upper() for p in periods}
    mentioned = re.findall(r'\bP\d{1,2}\b', narrative, re.IGNORECASE)
    return all(m.upper() in allowed for m in mentioned)


def _metrics_insight_fallback(entity_type: SummaryEntity, metrics: dict[str, Any]) -> str:
    tab_insights = metrics.get('tab_insights') or {}
    if tab_insights.get(entity_type):
        return tab_insights[entity_type]
    return build_tab_insights(metrics).get(entity_type, '')


def get_supervisor_summary(
    entity_type: SummaryEntity,
    filters: dict[str, Any],
    metrics: dict[str, Any] | None = None,
) -> dict[str, str]:
    prompt = build_supervisor_prompt(entity_type, filters, metrics)
    print(
        f'[supervisor] summary tab={entity_type} filters={json.dumps(filters)[:120]}',
        flush=True,
    )
    raw = _call_supervisor([{'role': 'user', 'content': prompt}], metrics)
    narrative = _extract_narrative(raw)

    if metrics and metrics.get('periods') and narrative:
        if not _narrative_uses_only_dashboard_periods(narrative, metrics['periods']):
            fallback = _metrics_insight_fallback(entity_type, metrics)
            if fallback:
                print(
                    f'[supervisor] {entity_type} cited periods outside dashboard '
                    f'({", ".join(metrics["periods"])}) — using metrics insight',
                    flush=True,
                )
                narrative = fallback

    if not narrative:
        raise RuntimeError('Supervisor returned empty narrative — retry shortly')

    return {'narrative': narrative, 'source': 'supervisor'}


def get_all_supervisor_summaries(
    filters: dict[str, Any],
    metrics: dict[str, Any] | None = None,
) -> dict[SummaryEntity, str]:
    tabs: list[SummaryEntity] = ['overview', 'category', 'line', 'dow', 'reason']
    print(f'[supervisor] PARALLEL mode — {len(tabs)} tab summaries (long_task={LONG_TASK})', flush=True)

    out: dict[SummaryEntity, str] = {}

    def fetch_tab(tab: SummaryEntity) -> tuple[SummaryEntity, str]:
        try:
            result = get_supervisor_summary(tab, filters, metrics)
            return tab, result['narrative']
        except Exception as e:
            print(f'[supervisor] {tab} failed: {e}', flush=True)
            return tab, ''

    with ThreadPoolExecutor(max_workers=len(tabs)) as pool:
        futures = [pool.submit(fetch_tab, tab) for tab in tabs]
        for fut in as_completed(futures):
            tab, narrative = fut.result()
            out[tab] = narrative

    return out
